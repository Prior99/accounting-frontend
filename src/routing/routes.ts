export const routeLogin = () => "/login";

export const routeDashboard = () => "/login";
