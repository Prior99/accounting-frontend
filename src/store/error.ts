import { observable, computed, action } from "mobx";

export interface DisplayableError {
    dismissed: boolean;
    message: string;
    origin: "api" | "internal";
    error?: Error;
}

export class ErrorStore {
    constructor() {
        this.apiError = this.apiError.bind(this);
        this.internalError = this.internalError.bind(this);
        this.dismissCurrent = this.dismissCurrent.bind(this);
    }

    @observable public errors: DisplayableError[] = [];

    @computed public get current() {
        return this.errors.find(error => !error.dismissed);
    }

    @action
    public dismissCurrent() {
        if (typeof this.current === "undefined") {
            return;
        }
        this.current.dismissed = true;
    }

    @action
    public apiError(message: string, error?: Error) {
        this.errors.push({ dismissed: false, message, origin: "api", error });
    }

    @action
    public internalError(message: string, error?: Error) {
        this.errors.push({ dismissed: false, message, origin: "internal", error });
    }
}
