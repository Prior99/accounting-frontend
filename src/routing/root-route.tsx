import * as React from "react";
import { Route, Redirect, browserHistory, IndexRoute } from "react-router";

import {
    PageLogin,
    PageDashboard,
} from "../pages";
import { AppContainer } from "../ui";
import { routeLogin } from "./routes";
import { Store } from "../store";

function requireLogin() {
    if (!Store.login.loggedIn) {
        browserHistory.replace(routeLogin());
    }
}

export const rootRoute = (
    <Route component={AppContainer}>
        <Redirect from="/" to="/login" />
        <Route path="login" component={PageLogin} />
        <Route path="dashboard" component={PageDashboard} onEnter={requireLogin}/>
    </Route>
);
