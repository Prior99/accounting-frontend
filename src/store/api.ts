import { observable, computed, action } from "mobx";
import { bind } from "bind-decorator";
import * as Api from "../../generated/api/api";
import * as isomorphicFetch from "isomorphic-fetch";
import { RequestStatus } from "../request-status";
import { History } from "history";
import * as routes from "../routing";
import { Store } from "./store";

declare var baseUrl: string;
const prefixedBaseUrl = `${window.location.protocol}//${baseUrl}`;

const softwareVersion = 1;
const localStorageIdentifier = "accounting-software-login";

interface LocalStorageApi {
    storageVersion: number;
    date: string;
    authToken: string;
}

interface ApiError {
    message: string;
}

interface Requests {
    [key: string]: {
        error?: ApiError;
        requestStatus: RequestStatus;
    };
}

export class ApiStore extends Store {
    @observable public authToken: string;
    @observable public requests: Requests = {};
    @observable public errors: ApiError[] = [];

    public auth = new Api.AuthApi(this.wrappedFetch, prefixedBaseUrl);
    public info = new Api.InfoApi(this.wrappedFetch, prefixedBaseUrl);
    public login = new Api.LoginApi(this.wrappedFetch, prefixedBaseUrl);
    public member = new Api.MemberApi(this.wrappedFetch, prefixedBaseUrl);
    public registration = new Api.RegistrationApi(this.wrappedFetch, prefixedBaseUrl);
    public teams = new Api.TeamsApi(this.wrappedFetch, prefixedBaseUrl);
    public timer = new Api.TimerApi(this.wrappedFetch, prefixedBaseUrl);
    public user = new Api.UserApi(this.wrappedFetch, prefixedBaseUrl);

    constructor(browserHistory: History) {
        super(undefined, browserHistory);
        this.load();
    }

    @computed
    public get loggedIn() {
        return typeof this.authToken !== "undefined";
    }

    @bind
    private wrappedFetch(url: string, options: any) {
        const headers = options.headers || new Headers();
        if (this.loggedIn) {
            headers.append("authorization", this.authToken);
        }
        return isomorphicFetch(url, { ...options, headers });
    }

    @bind @action
    public async call<T>(identifier: string, request: () => Promise<T>): Promise<T> {
        try {
            this.requests[identifier] = { requestStatus: RequestStatus.IN_PROGRESS };
            const response = await request();
            this.requests[identifier] = { requestStatus: RequestStatus.SUCCESS };
            return response;
        } catch (error) {
            this.requests[identifier] = { requestStatus: RequestStatus.FAIL, error };
            this.errors.push(error);
            throw error;
        }
    }

    @bind @action
    public async doLogin(email: string, password: string) {
        const body = { email, password };
        const response = await this.call("doLogin", () => this.login.v1AuthLoginPost({ body }));
        if (response) {
            const { token } = response;
            this.authToken = token;
            this.save();
            this.browserHistory.replace(routes.dashboard());
        }
    }

    @bind
    private save() {
        const deserialized: LocalStorageApi = {
            storageVersion: softwareVersion,
            date: new Date().toString(),
            authToken: this.authToken,
        };
        const serialized = JSON.stringify(deserialized);
        localStorage.setItem(localStorageIdentifier, serialized);
    }

    @bind
    private clearStorage() {
        localStorage.removeItem(localStorageIdentifier);
    }

    @bind
    private load() {
        const serialized = localStorage.getItem(localStorageIdentifier);
        if (serialized === null) { // tslint:disable-line
            return;
        }
        let deserialized: LocalStorageApi;
        try {
            deserialized = JSON.parse(serialized);
        } catch (err) {
            this.clearStorage();
            return;
        }
        if (deserialized.storageVersion !== softwareVersion) {
            this.clearStorage();
            return;
        }
        this.authToken = deserialized.authToken;
        this.browserHistory.replace(routes.dashboard());
    }

    public getRequestStatus(identifier: string): RequestStatus {
        const request = this.requests[identifier];
        if (typeof request === "undefined") {
            return RequestStatus.NONE;
        }
        return request.requestStatus;
    }

    public getError(identifier: string): ApiError {
        const request = this.requests[identifier];
        if (typeof request === "undefined") {
            return undefined;
        }
        return request.error;
    }

    @bind @action
    public doDismiss() {
        this.errors.pop();
    }

    @computed
    public get latestError() {
        return this.errors[this.errors.length - 1];
    }
}
