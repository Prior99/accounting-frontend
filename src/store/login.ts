import { observable, computed, action } from "mobx";
import { api } from "./api";
import { bind } from "bind-decorator";

export class LoginStore {
    @observable public authToken: string;
    @observable public failed = false;

    @computed
    public get loggedIn() {
        return typeof this.authToken !== "undefined";
    }

    @bind
    @action
    public async onLogin(email: string, password: string) {
        const result = await api("/auth/login", { email, password }, "POST", true);
        const { data } = result;
        if (result.okay && data.auth_token) {
            this.authToken = data.auth_token;
        } else {
            this.failed = true;
        }
    }
}
