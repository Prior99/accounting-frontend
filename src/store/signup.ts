import { observable, computed, action } from "mobx";
import { bind } from "bind-decorator";
import { ApiStore } from ".";
import { History } from "history";
import { Store } from "./store";

export class SignupStore extends Store {
    @observable
    public signupResult: Boolean;

    @bind @action
    public async doSignup(email: string, password: string) {
        const body = { email, password };
        const response = await this.api.call("doSignup", () => this.api.registration.v1AuthRegisterPost({ body }));
        if (response) {
            await this.api.doLogin(email, password);
            this.signupResult = true;
        } else {
            this.signupResult = false;
        }
    }
}
