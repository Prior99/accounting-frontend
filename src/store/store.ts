import { ApiStore } from "./api";
import { History } from "history";

export class Store {
    protected api: ApiStore;
    protected browserHistory: History;

    constructor(api: ApiStore, browserHistory: History) {
        this.api = api;
        this.browserHistory = browserHistory;
    }
}
