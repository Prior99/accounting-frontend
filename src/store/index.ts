import { LoginStore } from "./login";
import { SignupStore } from "./signup";
import { ErrorStore } from "./error";
export * from "./error";

export interface Store {
    login: LoginStore;
    error: ErrorStore;
    signup: SignupStore;
}

export let store: Store;

export function createStore(): Store {
    return {
        login: new LoginStore(),
        error: new ErrorStore(),
        signup: new SignupStore()
    };
}

export function resetStore() {
    store = createStore();
}

resetStore();

if (typeof window !== "undefined") {
    (window as any).store = store;
    (window as any).createStore = createStore;
    (window as any).resetStore = resetStore;
}
