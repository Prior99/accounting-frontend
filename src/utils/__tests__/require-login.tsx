import * as React from "react";
import * as routes from "routing";
import { requireLogin } from "..";
import { createRenderer, ShallowRenderer } from "react-test-renderer/shallow";
import { ApiStore } from "store";
import { Provider } from "mobx-react";

class SomeComponent extends React.PureComponent<{}> {
    public render() {
        return (
            <div>Some Component</div>
        );
    }
}

let DecoratedComponent: React.ComponentClass<{}>;
let renderer: ShallowRenderer;
let apiStore: ApiStore;

beforeEach(() => {
    renderer = createRenderer();
    DecoratedComponent = requireLogin(SomeComponent);
    apiStore = new ApiStore(undefined);
});

test("`requireLogin()` redirects to the `login()` route if not logged in", () => {
    apiStore.authToken = undefined;
    const current = renderer.render(
        <Provider api={apiStore}>
            <DecoratedComponent />
        </Provider>,
    );
    expect(current).toMatchSnapshot();
});

test("`requireLogin()` does not redirect if logged in", () => {
    apiStore.authToken = "some-auth-token";
    const current = renderer.render(
        <Provider api={apiStore}>
            <DecoratedComponent />
        </Provider>,
    );
    expect(current).toMatchSnapshot();
});
