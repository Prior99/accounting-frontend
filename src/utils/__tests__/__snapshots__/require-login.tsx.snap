// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`requireLogin()\` does not redirect if logged in 1`] = `<inject-RequireLogin-with-api />`;

exports[`\`requireLogin()\` redirects to the \`login()\` route if not logged in 1`] = `<inject-RequireLogin-with-api />`;
